### PR TITLE
Pre-create DNS records

### PR DESCRIPTION
**Pull request by justinsb**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1006)

### CODE REVIEW

File: upup/pkg/fi/cloudup/apply_cluster.go [hunk 1]
```diff
@@ -643,6 +649,79 @@ func validateDNS(cluster *api.Cluster, cloud fi.Cloud) error {
 		glog.V(2).Infof("Found NS records for %q: %v", dnsName, hosts)
 	}
 
+	// TODO: Move to update
+	if os.Getenv("KOPS_FF_SKIP_DNS_PRECREATE") == "" {
```
Comment (kris-nova) [inline upup/pkg/fi/cloudup/apply_cluster.go]: I know it's a WIP - so don't mind me snooping :) 

Are we keeping the `$ENVAR`'s in kops? If so just curious if we can/should start a manifest of all of them, with their behavior?

Comment (justinsb) [inline upup/pkg/fi/cloudup/apply_cluster.go]: Good news ... I added a real feature flag implementation.  (Still pretty experimental)  But I'm going to update this PR to use that...

**Comment by justinsb** [comment]
Fixes #928

**Comment by justinsb** [comment]
Revised & ready for review I think.

This can be turned _off_ using `export KOPS_FEATURE_FLAG=-DNSPreCreate`

**Comment by chrislovecnm** [comment]
CI is busted ;)

File: pkg/featureflag/featureflag.go [hunk 2]
```diff
@@ -22,16 +22,35 @@ import (
 	"sync"
 )
 
-var PreviewPrivateDNS = New("PreviewPrivateDNS")
+func Bool(b bool) *bool {
+	return &b
+}
+
+// PreviewPrivateDNS turns on the preview of the private hosted zone support
```
Comment (chrislovecnm) [inline pkg/featureflag/featureflag.go]: If we are doing sentences do we need periods?

Comment (chrislovecnm) [inline pkg/featureflag/featureflag.go]: Literally do not know what our style is going to be like

**Comment by zmerlynn** [comment]
@chrislovecnm: Ha, for once it's the Travis CI. I was going to look into the PR builder today (it seems to be faring much worse in the timeout department than the main AWS run).

File: upup/pkg/fi/cloudup/apply_cluster.go [hunk 3]
```diff
@@ -548,6 +546,14 @@ func (c *ApplyClusterCmd) Run() error {
 	if err != nil {
 		return fmt.Errorf("error running tasks: %v", err)
 	}
+
+	if !dryRun {
+		err = precreateDNS(cluster, cloud)
+		if err != nil {
```
Comment (zmerlynn) [inline upup/pkg/fi/cloudup/apply_cluster.go]: If you have a free scope block, you might as well `err := ` here. For bonus points, `if err := foo; err != nil {}` (but I recognize it's not prevailing `kops` style)

File: upup/pkg/fi/cloudup/dns.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+	"github.com/golang/glog"
+	api "k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/featureflag"
+	"k8s.io/kops/pkg/model"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider/rrstype"
+	"net"
+	"os"
+	"strings"
```
Comment (zmerlynn) [inline upup/pkg/fi/cloudup/dns.go]: Out of curiousity, does your `gofmt` behave differently than mine? I find when I edit `kops` files, the first time I save it, it tends to sort the imports to two blocks, base imports and external imports, each alphabetical.

File: upup/pkg/fi/cloudup/dns.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+		return nil, fmt.Errorf("found multiple DNS Zones matching %q", cluster.Spec.DNSZone)
+	}
+
+	zone := matches[0]
+	return zone, nil
+}
+
+func validateDNS(cluster *api.Cluster, cloud fi.Cloud) error {
+	kopsModelContext := &model.KopsModelContext{
+		//Region: cloud.Region(),
```
Comment (zmerlynn) [inline upup/pkg/fi/cloudup/dns.go]: Is the dead code valuable?

File: upup/pkg/fi/cloudup/dns.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+	dnsName := strings.TrimSuffix(zone.Name(), ".")
+
+	glog.V(2).Infof("Doing DNS lookup to verify NS records for %q", dnsName)
+	ns, err := net.LookupNS(dnsName)
+	if err != nil {
+		return fmt.Errorf("error doing DNS lookup for NS records for %q: %v", dnsName, err)
+	}
+
+	if len(ns) == 0 {
+		if os.Getenv("DNS_IGNORE_NS_CHECK") == "" {
```
Comment (zmerlynn) [inline upup/pkg/fi/cloudup/dns.go]: Eventually we should have a conversation about env vs flags. Env variables are kind of frustrating action-at-a-distance - I prefer to keep it encapsulated in flags if we can. (Yes, I realize this is old code you're moving, but it inspired a comment 😄  )

File: pkg/featureflag/featureflag.go (hunk 2 above)
Comment (zmerlynn) [inline pkg/featureflag/featureflag.go]: All documentation for exported pieces should be readable as English documentation - it makes godoc from the source readable. We're not great at enforcing this or the requirement to document exported pieces, though.

**Comment by zmerlynn** [comment]
Oh, yeah, and fix the tests. It looks like something is up with `MockRoute53`. :)
```
--- FAIL: TestMinimal (1.76s)
panic: MockRoute53 function not implemented [recovered]
	panic: MockRoute53 function not implemented
```

**Comment by zmerlynn** [comment]
Also, I'm excited for this change. The PR builder shows only a 5m bringup, which is well faster than I've been seeing:
```
I1227 19:39:30.078] Cluster is starting.  It should be ready in a few minutes.
[...]
W1227 19:44:31.830] 2016/12/27 19:44:31 e2e.go:760: Cluster nodes:
```

File: pkg/featureflag/featureflag.go (hunk 2 above)
Comment (justinsb) [inline pkg/featureflag/featureflag.go]: Added full stops at the end of the sentences

File: upup/pkg/fi/cloudup/apply_cluster.go (hunk 3 above)
Comment (justinsb) [inline upup/pkg/fi/cloudup/apply_cluster.go]: Fixed

File: upup/pkg/fi/cloudup/dns.go (hunk 4 above)
Comment (justinsb) [inline upup/pkg/fi/cloudup/dns.go]: Have you aliased `goimports` maybe?

I tend to run `make gofmt`, which hopefully works the same for everyone.... though if you have aliased goimports I guess it won't

File: upup/pkg/fi/cloudup/dns.go (hunk 5 above)
Comment (justinsb) [inline upup/pkg/fi/cloudup/dns.go]: Added a comment to better clarify the point I was trying to make - that it isn't a "full" KopsModelContext

File: upup/pkg/fi/cloudup/dns.go (hunk 4 above)
Comment (zmerlynn) [inline upup/pkg/fi/cloudup/dns.go]: Oh, yes. I forgot I did that, everything has been working too well. :)

File: upup/pkg/fi/cloudup/dns.go (hunk 6 above)
Comment (justinsb) [inline upup/pkg/fi/cloudup/dns.go]: Agree... I added FeatureFlags to hopefully normalize our use of env vars.

I added this godoc to featureflags.go

```
// Package featureflag implements simple feature-flagging.
// Feature flags can become an anti-pattern if abused.
// We should try to use them for two use-cases:
// * `Preview` feature flags enable a piece of functionality we haven't yet fully baked.  The user needs to 'opt-in'.
//   We expect these flags to be removed at some time.  Normally these will default to false.
// * Escape-hatch feature flags turn off a default that we consider risky (e.g. pre-creating DNS records).
//   This lets us ship a behaviour, and if we encounter unusual circumstances in the field, we can
//   allow the user to turn the behaviour off.  Normally these will default to true.
```

**Comment by justinsb** [comment]
Just fixing up mockroute53 - I need to implement the required functions.

(mockroute53 / mockec2 is a "fake" in-memory AWS API implementation.  Very limited right now).

**Comment by justinsb** [comment]
Looks like e2e was slow, but bringup was still fast.  e2e was slow because of a PetSet test timeout, it looks like.

**Comment by zmerlynn** [comment]
Probably the same issue as `kops-aws` is hitting right now (it's been stuck making no progress for over an hour).

**Pull request merged**